[Concurrency][Embedded] Make sure Embedded Swift links Impl functions.

al45tair opened the pull request:
Teach the Embedded Swift compiler to include the Impl functions for the functions that provide Concurrency hooks.

al45tair committed:
[Concurrency][Embedded] Make sure Embedded Swift links Impl functions.

Teach the Embedded Swift compiler to include the Impl functions for the
functions that provide Concurrency hooks.

al45tair commented:
@swift-ci Please smoke test

kubamracek commented:
Makes sense, but could we also get some test that shows the problem that this solves?

al45tair committed:
[Concurrency][Embedded] Add some extra Concurrency functions.

We need to explicitly link `swift_asyncMainDrainQueueImpl` and
`swift_createDefaultExecutors` since those two do not have Concurrency
hooks and so aren't linked automatically by the macro-processing code.

MaxDesiatov committed:
Link.cpp: load `MainActor: Actor` WT for Embedded Swift

MaxDesiatov committed:
Add `linkWitnessTable` to `SILModule`, use in `Link.cpp`

MaxDesiatov committed:
Enable more `MainActor` code paths for Embedded Swift

MaxDesiatov commented:
> Makes sense, but could we also get some test that shows the problem that this solves?

It would allow us to enable this test https://github.com/swiftlang/swift/pull/83236, but I suggest we fix `swift_deletedAsyncMethodError` issue that also blocks that test in a separate PR.

MaxDesiatov committed:
Executor.swift: only enable `MainActor` in Embedded Swift for WASI
